<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Nextflow 2 &mdash; Courses@CRG: Reproducible research and data analysis using Containers &amp; Nextflow 2022  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="canonical" href="https://biocorecrg.github.io/CoursesCRG_Containers_Nextflow_May_2022/nextflow_2.html" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Nextflow 3" href="nextflow_3.html" />
    <link rel="prev" title="Nextflow 1" href="nextflow_1.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> Courses@CRG: Reproducible research and data analysis using Containers & Nextflow 2022
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="about.html">About the course</a></li>
<li class="toctree-l1"><a class="reference internal" href="nextflow_1.html">Nextflow 1</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Nextflow 2</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#more-complex-scripts">More complex scripts</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#exercise">Exercise</a></li>
<li class="toctree-l3"><a class="reference internal" href="#resuming-a-pipeline">Resuming a pipeline</a></li>
<li class="toctree-l3"><a class="reference internal" href="#directives">Directives</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id1">EXERCISE</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#decoupling-resources-parameters-and-nextflow-script">Decoupling resources, parameters and nextflow script</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#publishing-final-results">Publishing final results</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#adding-help-section-to-a-pipeline">Adding help section to a pipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id2">EXERCISE</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="nextflow_3.html">Nextflow 3</a></li>
<li class="toctree-l1"><a class="reference internal" href="nextflow_4.html">Nextflow 4</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Courses@CRG: Reproducible research and data analysis using Containers & Nextflow 2022</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Nextflow 2</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/nextflow_2.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="nextflow-2">
<span id="nextflow-2-page"></span><h1>Nextflow 2<a class="headerlink" href="#nextflow-2" title="Permalink to this heading"></a></h1>
<section id="more-complex-scripts">
<h2>More complex scripts<a class="headerlink" href="#more-complex-scripts" title="Permalink to this heading"></a></h2>
<p>We can feed the channel generated by a process to another process in the workflow definition. The variables used by AWK need to be escaped, otherwise they will be considered as proper Nextflow variables and thus produce an error. Every special character, e.g., <strong>$</strong>, needs to be escaped (<strong>$</strong>). It can be tedeous when writing long one liners; therefore, it is recommended to make a small shell script and call it as an executable. It has to be placed in a folder named <strong>bin</strong> inside the pipeline folder to be automatically considered from Nextflow as a tool in the path. This script can be seen at <strong>/test1/test1.nf</strong></p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="cp">#!/usr/bin/env nextflow</span><span class="w"></span>


<span class="cm">/* </span>
<span class="cm"> * This code enables the new dsl of Nextflow. </span>
<span class="cm"> */</span><span class="w"></span>
<span class="w"> </span>
<span class="n">nextflow</span><span class="o">.</span><span class="na">enable</span><span class="o">.</span><span class="na">dsl</span><span class="o">=</span><span class="mi">2</span><span class="w"></span>

<span class="cm">/* </span>
<span class="cm"> * HERE YOU HAVE THE COMMMENTS</span>
<span class="cm"> * NextFlow example from their website </span>
<span class="cm"> */</span><span class="w"></span>
<span class="w"> </span>
<span class="c1">// this can be overridden by using --inputfile OTHERFILENAME</span><span class="w"></span>
<span class="n">params</span><span class="o">.</span><span class="na">inputfile</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;$baseDir/../../testdata/test.fa&quot;</span><span class="w">	</span>

<span class="c1">// the &quot;file method&quot; returns a file system object given a file path string  </span><span class="w"></span>
<span class="n">sequences_file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">inputfile</span><span class="o">)</span><span class="w">				</span>

<span class="c1">// check if the file exists</span><span class="w"></span>
<span class="k">if</span><span class="o">(</span><span class="w"> </span><span class="o">!</span><span class="n">sequences_file</span><span class="o">.</span><span class="na">exists</span><span class="o">()</span><span class="w"> </span><span class="o">)</span><span class="w"> </span><span class="n">exit</span><span class="w"> </span><span class="mi">1</span><span class="o">,</span><span class="w"> </span><span class="s2">&quot;Missing genome file: ${sequences_file}&quot;</span><span class="w"> </span>


<span class="cm">/*</span>
<span class="cm"> * split a fasta file in multiple files</span>
<span class="cm"> */</span><span class="w"></span>
<span class="w"> </span>
<span class="n">process</span><span class="w"> </span><span class="n">splitSequences</span><span class="w"> </span><span class="o">{</span><span class="w"></span>

<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">sequencesFile</span><span class="w"> </span><span class="c1">// nextflow creates links to the original files in a temporary folder</span><span class="w"></span>
<span class="w"> </span>
<span class="w">    </span><span class="nl">output:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="o">(</span><span class="s1">&#39;seq_*&#39;</span><span class="o">)</span><span class="w">    </span><span class="c1">// send output files to a new output channel (in this case is a collection)</span><span class="w"></span>
<span class="w"> </span>
<span class="w">    </span><span class="c1">// awk command for splitting a fasta files in multiple files</span><span class="w"></span>
<span class="w">    </span>
<span class="w">    </span><span class="nl">script:</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    awk &#39;/^&gt;/{f=&quot;seq_&quot;++d} {print &gt; f}&#39; &lt; ${sequencesFile}</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"> </span>
<span class="o">}</span><span class="w"></span>


<span class="cm">/*</span>
<span class="cm"> * Simple reverse the sequences</span>
<span class="cm"> */</span><span class="w"></span>
<span class="w"> </span>
<span class="n">process</span><span class="w"> </span><span class="n">reverseSequence</span><span class="w"> </span><span class="o">{</span><span class="w"></span>

<span class="w">    </span><span class="c1">// during the execution prints the indicated variable for follow-up</span><span class="w"></span>
<span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="s2">&quot;${seq}&quot;</span><span class="w"> </span><span class="o">}</span><span class="w">  					</span>

<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">seq</span><span class="w"> </span>

<span class="w">    </span><span class="nl">output:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;all.rev&quot;</span><span class="w"> </span>
<span class="w"> </span>
<span class="w">    </span><span class="nl">script:</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    cat ${seq} | awk &#39;{if (\$1~&quot;&gt;&quot;) {print \$0} else system(&quot;echo &quot; \$0 &quot; |rev&quot;)}&#39; &gt; all.rev</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="n">workflow</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">splitted_seq</span><span class="w">	</span><span class="o">=</span><span class="w"> </span><span class="n">splitSequences</span><span class="o">(</span><span class="n">sequences_file</span><span class="o">)</span><span class="w"></span>
<span class="w">    </span>
<span class="w">    </span><span class="c1">// Here you have the output channel as a collection</span><span class="w"></span>
<span class="w">    </span><span class="n">splitted_seq</span><span class="o">.</span><span class="na">view</span><span class="o">()</span><span class="w"></span>
<span class="w">    </span>
<span class="w">    </span><span class="c1">// Here you have the same channel reshaped to send separately each value </span><span class="w"></span>
<span class="w">    </span><span class="n">splitted_seq</span><span class="o">.</span><span class="na">flatten</span><span class="o">().</span><span class="na">view</span><span class="o">()</span><span class="w"></span>
<span class="w">    </span>
<span class="w">    </span><span class="c1">// DLS2 allows you to reuse the channels! In past you had to create many identical</span><span class="w"></span>
<span class="w">    </span><span class="c1">// channels for sending the same kind of data to different processes</span><span class="w"></span>
<span class="w">    </span>
<span class="w">    </span><span class="n">rev_single_seq</span><span class="w">	</span><span class="o">=</span><span class="w"> </span><span class="n">reverseSequence</span><span class="o">(</span><span class="n">splitted_seq</span><span class="o">)</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
<p>Here we have two simple processes:</p>
<ul class="simple">
<li><p>the former splits the input fasta file into <strong>single sequences</strong>.</p></li>
<li><p>the latter is able to <strong>reverse the position of the sequences</strong>.</p></li>
</ul>
<p>The input path is fed as a parameter using the script parameters <strong>${seq}</strong></p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="n">params</span><span class="o">.</span><span class="na">inputfile</span><span class="w"></span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The file “test.fa” is available in the <a class="reference external" href="https://github.com/biocorecrg/CRG_Nextflow_Jun_2022/tree/main/testdata">github repository of the course</a></p>
</div>
<p>This value can be overridden when calling the script:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nextflow run test1.nf --inputfile another_input.fa</span>
</pre></div>
</div>
<p>The workflow part connects the two processes so that the output of the first process becomes an input of the second process.</p>
<p>During the execution, Nextflow creates a number of temporary folders and also a soft link to the original input file. It will then store output files locally.</p>
<p>The output file is then <em>linked</em> in other folders for being used as input from other processes.
This avoids clashes, and each process is isolated from the other.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nextflow run test1.nf -bg</span>

<span class="go">N E X T F L O W  ~  version 20.07.1</span>
<span class="go">Launching `test1.nf` [sad_newton] - revision: 82e66714e4</span>
<span class="go">[09/53e071] Submitted process &gt; splitSequences</span>
<span class="go">[/home/ec2-user/git/CRG_Nextflow_Jun_2022/nextflow/nextflow/work/09/53e071d286ed66f4020869c8977b59/seq_1, /home/ec2-user/git/CRG_Nextflow_Jun_2022/nextflow/nextflow/work/09/53e071d286ed66f4020869c8977b59/seq_2, /home/ec2-user/git/CRG_Nextflow_Jun_2022/nextflow/nextflow/work/09/53e071d286ed66f4020869c8977b59/seq_3]</span>
<span class="go">/home/ec2-user/git/CRG_Nextflow_Jun_2022/nextflow/nextflow/work/09/53e071d286ed66f4020869c8977b59/seq_1</span>
<span class="go">/home/ec2-user/git/CRG_Nextflow_Jun_2022/nextflow/nextflow/work/09/53e071d286ed66f4020869c8977b59/seq_2</span>
<span class="go">/home/ec2-user/git/CRG_Nextflow_Jun_2022/nextflow/nextflow/work/09/53e071d286ed66f4020869c8977b59/seq_3</span>
<span class="go">[fe/0a8640] Submitted process &gt; reverseSequence ([seq_1, seq_2, seq_3])</span>
</pre></div>
</div>
<p>We can inspect the content of <cite>work/09/53e071*</cite> generated by the process <strong>splitSequences</strong>:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">ls -l work/09/53e071*</span>
<span class="go">total 24</span>
<span class="go">-rw-r--r--  1 lcozzuto  staff  29 Oct  8 19:16 seq_1</span>
<span class="go">-rw-r--r--  1 lcozzuto  staff  33 Oct  8 19:16 seq_2</span>
<span class="go">-rw-r--r--  1 lcozzuto  staff  27 Oct  8 19:16 seq_3</span>
<span class="go">lrwxr-xr-x  1 lcozzuto  staff  69 Oct  8 19:16 test.fa -&gt; /home/ec2-user/git/CRG_Nextflow_Jun_2022/nextflow/nextflow/testdata/test.fa</span>
</pre></div>
</div>
<p>File <cite>test.fa</cite> is a <em>soft link</em> to the original input.</p>
<p>If we inspect <cite>work/fe/0a8640*</cite> that is generated by the process <strong>reverseSequence</strong>, we see that the files generated by <strong>splitSequences</strong> are now linked as input.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">ls -l work/fe/0a8640*</span>

<span class="go">total 8</span>
<span class="go">-rw-r--r--  1 lcozzuto  staff  89 Oct  8 19:16 all.rev</span>
<span class="go">lrwxr-xr-x  1 lcozzuto  staff  97 Oct  8 19:16 seq_1 -&gt; /home/ec2-user/git/CRG_Nextflow_Jun_2022/nextflow/nextflow/work/09/53e071d286ed66f4020869c8977b59/seq_1</span>
<span class="go">lrwxr-xr-x  1 lcozzuto  staff  97 Oct  8 19:16 seq_2 -&gt; /home/ec2-user/git/CRG_Nextflow_Jun_2022/nextflow/nextflow/work/09/53e071d286ed66f4020869c8977b59/seq_2</span>
<span class="go">lrwxr-xr-x  1 lcozzuto  staff  97 Oct  8 19:16 seq_3 -&gt; /home/ec2-user/git/CRG_Nextflow_Jun_2022/nextflow/nextflow/work/09/53e071d286ed66f4020869c8977b59/seq_3</span>
</pre></div>
</div>
<p>At this point we can make two different workflows to demonstrate how the new DSL allows reusing the code (script name <strong>test1_b.nf</strong>).</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="cp">#!/usr/bin/env nextflow</span><span class="w"></span>

<span class="n">nextflow</span><span class="o">.</span><span class="na">enable</span><span class="o">.</span><span class="na">dsl</span><span class="o">=</span><span class="mi">2</span><span class="w"></span>

<span class="c1">// this can be overridden by using --inputfile OTHERFILENAME</span><span class="w"></span>
<span class="n">params</span><span class="o">.</span><span class="na">inputfile</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;$baseDir/../../testdata/test.fa&quot;</span><span class="w"></span>

<span class="c1">// the &quot;file method&quot; returns a file system object given a file path string</span><span class="w"></span>
<span class="n">sequences_file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">inputfile</span><span class="o">)</span><span class="w"></span>

<span class="c1">// check if the file exists</span><span class="w"></span>
<span class="k">if</span><span class="o">(</span><span class="w"> </span><span class="o">!</span><span class="n">sequences_file</span><span class="o">.</span><span class="na">exists</span><span class="o">()</span><span class="w"> </span><span class="o">)</span><span class="w"> </span><span class="n">exit</span><span class="w"> </span><span class="mi">1</span><span class="o">,</span><span class="w"> </span><span class="s2">&quot;Missing genome file: ${sequences_file}&quot;</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Process 1 for splitting a fasta file in multiple files</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="n">splitSequences</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">sequencesFile</span><span class="w"></span>

<span class="w">    </span><span class="nl">output:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="o">(</span><span class="s1">&#39;seq_*&#39;</span><span class="o">)</span><span class="w"></span>

<span class="w">    </span><span class="c1">// simple awk command</span><span class="w"></span>
<span class="w">    </span><span class="nl">script:</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    awk &#39;/^&gt;/{f=&quot;seq_&quot;++d} {print &gt; f}&#39; &lt; ${sequencesFile}</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Process 2 for reversing the sequences</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="n">reverseSequence</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="s2">&quot;${seq}&quot;</span><span class="w"> </span><span class="o">}</span><span class="w"></span>

<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">seq</span><span class="w"></span>

<span class="w">    </span><span class="nl">output:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;all.rev&quot;</span><span class="w"></span>

<span class="w">    </span><span class="nl">script:</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    	cat ${seq} | awk &#39;{if (\$1~&quot;&gt;&quot;) {print \$0} else system(&quot;echo &quot; \$0 &quot; |rev&quot;)}&#39; &gt; all.rev</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="hll"><span class="n">workflow</span><span class="w"> </span><span class="n">flow1</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
</span><span class="hll"><span class="w">    </span><span class="nl">take:</span><span class="w"> </span><span class="n">sequences</span><span class="w"></span>
</span><span class="hll">
</span><span class="hll"><span class="w">    </span><span class="nl">main:</span><span class="w"></span>
</span><span class="hll"><span class="w">    </span><span class="n">splitted_seq</span><span class="w">        </span><span class="o">=</span><span class="w"> </span><span class="n">splitSequences</span><span class="o">(</span><span class="n">sequences</span><span class="o">)</span><span class="w"></span>
</span><span class="hll"><span class="w">    </span><span class="n">rev_single_seq</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="n">reverseSequence</span><span class="o">(</span><span class="n">splitted_seq</span><span class="o">)</span><span class="w"></span>
</span><span class="hll"><span class="o">}</span><span class="w"></span>
</span><span class="hll">
</span><span class="hll"><span class="n">workflow</span><span class="w"> </span><span class="n">flow2</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
</span><span class="hll"><span class="w">    </span><span class="nl">take:</span><span class="w"> </span><span class="n">sequences</span><span class="w"></span>
</span><span class="hll">
</span><span class="hll"><span class="w">    </span><span class="nl">main:</span><span class="w"></span>
</span><span class="hll"><span class="w">    </span><span class="n">splitted_seq</span><span class="w">        </span><span class="o">=</span><span class="w"> </span><span class="n">splitSequences</span><span class="o">(</span><span class="n">sequences</span><span class="o">).</span><span class="na">flatten</span><span class="o">()</span><span class="w"></span>
</span><span class="hll"><span class="w">    </span><span class="n">rev_single_seq</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="n">reverseSequence</span><span class="o">(</span><span class="n">splitted_seq</span><span class="o">)</span><span class="w"></span>
</span><span class="hll"><span class="o">}</span><span class="w"></span>
</span>
<span class="n">workflow</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">   </span><span class="n">flow1</span><span class="o">(</span><span class="n">sequences_file</span><span class="o">)</span><span class="w"></span>
<span class="w">   </span><span class="n">flow2</span><span class="o">(</span><span class="n">sequences_file</span><span class="o">)</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
<p>The first workflow will just run like the previous script, while the second will “flatten” the output of the first process and will launch the second process on each single sequence.</p>
<p>The <strong>reverseSequence</strong> process of the second workflow will run in parallel if you have enough processors, or if you are running the script in a cluster environment, with a scheduler supported by Nextflow.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nextflow run test1_b.nf -bg</span>

<span class="gp">C02WX1XFHV2Q:nextflow lcozzuto$ </span>N E X T F L O W  ~  version <span class="m">20</span>.07.1
<span class="go">Launching `test1.nf` [insane_plateau] - revision: d33befe154</span>
<span class="go">[bd/f4e9a6] Submitted process &gt; flow1:splitSequences</span>
<span class="go">[37/d790ab] Submitted process &gt; flow2:splitSequences</span>
<span class="go">[33/a6fc72] Submitted process &gt; flow1:reverseSequence ([seq_1, seq_2, seq_3])</span>
<span class="go">[87/54bfe8] Submitted process &gt; flow2:reverseSequence (seq_2)</span>
<span class="go">[45/86dd83] Submitted process &gt; flow2:reverseSequence (seq_1)</span>
<span class="go">[93/c7b1c6] Submitted process &gt; flow2:reverseSequence (seq_3)</span>
</pre></div>
</div>
<section id="exercise">
<h3>Exercise<a class="headerlink" href="#exercise" title="Permalink to this heading"></a></h3>
<p>Optimize the previus pipeline to avoid running the process <strong>splitSequences</strong> twice.</p>
<details>
<summary><a>Solution</a></summary><div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="cp">#!/usr/bin/env nextflow</span><span class="w"></span>

<span class="n">nextflow</span><span class="o">.</span><span class="na">enable</span><span class="o">.</span><span class="na">dsl</span><span class="o">=</span><span class="mi">2</span><span class="w"></span>

<span class="c1">// this can be overridden by using --inputfile OTHERFILENAME</span><span class="w"></span>
<span class="n">params</span><span class="o">.</span><span class="na">inputfile</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;$baseDir/../../testdata/test.fa&quot;</span><span class="w"></span>

<span class="c1">// the &quot;file method&quot; returns a file system object given a file path string</span><span class="w"></span>
<span class="n">sequences_file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">inputfile</span><span class="o">)</span><span class="w"></span>

<span class="c1">// check if the file exists</span><span class="w"></span>
<span class="k">if</span><span class="o">(</span><span class="w"> </span><span class="o">!</span><span class="n">sequences_file</span><span class="o">.</span><span class="na">exists</span><span class="o">()</span><span class="w"> </span><span class="o">)</span><span class="w"> </span><span class="n">exit</span><span class="w"> </span><span class="mi">1</span><span class="o">,</span><span class="w"> </span><span class="s2">&quot;Missing genome file: ${sequences_file}&quot;</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Process 1 for splitting a fasta file in multiple files</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="n">splitSequences</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">sequencesFile</span><span class="w"></span>

<span class="w">    </span><span class="nl">output:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="o">(</span><span class="s1">&#39;seq_*&#39;</span><span class="o">)</span><span class="w"></span>

<span class="w">    </span><span class="c1">// simple awk command</span><span class="w"></span>
<span class="w">    </span><span class="nl">script:</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    awk &#39;/^&gt;/{f=&quot;seq_&quot;++d} {print &gt; f}&#39; &lt; ${sequencesFile}</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Process 2 for reversing the sequences</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="n">reverseSequence</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="s2">&quot;${seq}&quot;</span><span class="w"> </span><span class="o">}</span><span class="w"></span>

<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">seq</span><span class="w"></span>

<span class="w">    </span><span class="nl">output:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;all.rev&quot;</span><span class="w"></span>

<span class="w">    </span><span class="nl">script:</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    	cat ${seq} | awk &#39;{if (\$1~&quot;&gt;&quot;) {print \$0} else system(&quot;echo &quot; \$0 &quot; |rev&quot;)}&#39; &gt; all.rev</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="n">workflow</span><span class="w"> </span><span class="n">flow1</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="nl">take:</span><span class="w"> </span><span class="n">splitted_seq</span><span class="w"></span>

<span class="w">    </span><span class="nl">main:</span><span class="w"></span>
<span class="w">    </span><span class="n">rev_single_seq</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="n">reverseSequence</span><span class="o">(</span><span class="n">splitted_seq</span><span class="o">)</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="n">workflow</span><span class="w"> </span><span class="n">flow2</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="nl">take:</span><span class="w"> </span><span class="n">splitted_seq</span><span class="w"></span>

<span class="w">    </span><span class="nl">main:</span><span class="w"></span>
<span class="w">    </span><span class="n">rev_single_seq</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="n">reverseSequence</span><span class="o">(</span><span class="n">splitted_seq</span><span class="o">.</span><span class="na">flatten</span><span class="o">())</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="n">workflow</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">   </span><span class="n">splitted_seq</span><span class="w">        </span><span class="o">=</span><span class="w"> </span><span class="n">splitSequences</span><span class="o">(</span><span class="n">sequences_file</span><span class="o">)</span><span class="w"></span>
<span class="w">   </span><span class="n">flow1</span><span class="o">(</span><span class="n">splitted_seq</span><span class="o">)</span><span class="w"></span>
<span class="w">   </span><span class="n">flow2</span><span class="o">(</span><span class="n">splitted_seq</span><span class="o">)</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
</details><div class="line-block">
<div class="line"><br /></div>
<div class="line"><br /></div>
</div>
</section>
<section id="resuming-a-pipeline">
<h3>Resuming a pipeline<a class="headerlink" href="#resuming-a-pipeline" title="Permalink to this heading"></a></h3>
<p>You can resume the execution after the code modification using the parameter <strong>-resume</strong>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nextflow run test1.nf -bg -resume</span>

<span class="go">N E X T F L O W  ~  version 20.07.1</span>
<span class="go">Launching `test1.nf` [determined_celsius] - revision: eaf5b4d673</span>
<span class="go">[bd/f4e9a6] Cached process &gt; flow1:splitSequences</span>
<span class="go">[37/d790ab] Cached process &gt; flow2:splitSequences</span>
<span class="go">[93/c7b1c6] Cached process &gt; flow2:reverseSequence (seq_3)</span>
<span class="go">[45/86dd83] Cached process &gt; flow2:reverseSequence (seq_1)</span>
<span class="go">[87/54bfe8] Cached process &gt; flow2:reverseSequence (seq_2)</span>
<span class="go">[33/a6fc72] Cached process &gt; flow1:reverseSequence ([seq_1, seq_2, seq_3])</span>
<span class="go">/home/ec2-user/git/CRG_Nextflow_Jun_2022/nextflow/nextflow/work/33/a6fc72786d042cacf733034d501691/all.rev</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p><strong>IMPORTANT: Nextflow parameters are provided using one hyphen</strong> (<cite>-resume</cite>) <strong>while a pipeline parameters, two hyphens</strong> (<cite>--inputfile</cite>).</p>
</div>
<p>Sometimes you might want to resume a previous run of your pipeline.</p>
<p>To do so you need to extract the job id of that run. You can do this by using the command <cite>nextflow log</cite>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nextflow log</span>
<span class="go">TIMESTAMP               DURATION        RUN NAME                STATUS  REVISION ID     SESSION ID                              COMMAND</span>
<span class="go">2020-10-06 14:49:09     2s              agitated_avogadro       OK      61a595c5bf      4a7a8a4b-9bdb-4b15-9cc6-1b2cabe9a938    nextflow run test1.nf</span>
<span class="go">2020-10-08 19:14:38     2.8s            sick_edison             OK      82e66714e4      4fabb863-2038-47b4-bac0-19e71f93f284    nextflow run test1.nf -bg</span>
<span class="go">2020-10-08 19:16:03     3s              sad_newton              OK      82e66714e4      2d13e9f8-1ba6-422d-9087-5c6c9731a795    nextflow run test1.nf -bg</span>
<span class="go">2020-10-08 19:30:59     2.3s            disturbed_wozniak       OK      d33befe154      0a19b60d-d5fe-4a26-9e01-7a63d0a1d300    nextflow run test1.nf -bg</span>
<span class="go">2020-10-08 19:35:52     2.5s            insane_plateau          OK      d33befe154      b359f32c-254f-4271-95bb-6a91b281dc6d    nextflow run test1.nf -bg</span>
<span class="go">2020-10-08 19:56:30     2.8s            determined_celsius      OK      eaf5b4d673      b359f32c-254f-4271-95bb-6a91b281dc6d    nextflow run test1.nf -bg -resume</span>
</pre></div>
</div>
<p>You can then resume the state of your execution using the <strong>SESSION ID</strong>:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nextflow run -resume 0a19b60d-d5fe-4a26-9e01-7a63d0a1d300 test1.nf</span>
</pre></div>
</div>
<p>Nextflow’s cache can be disabled for a specific process by setting the directive <strong>cache</strong> to <strong>false</strong>. You can also choose among the three caching methods:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="n">cache</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="kc">true</span><span class="w"> </span><span class="c1">// (default) Cache keys are created indexing input files meta-data information (name, size and last update timestamp attributes).</span><span class="w"></span>

<span class="n">cache</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;deep&#39;</span><span class="w"> </span><span class="c1">// Cache keys are created indexing input files content.</span><span class="w"></span>

<span class="n">cache</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;lenient&#39;</span><span class="w"> </span><span class="c1">// (Best in HPC and shared file systems) Cache keys are created indexing input files path and size attributes</span><span class="w"></span>
</pre></div>
</div>
<p><strong>IMPORTANT: On some shared file systems you might have inconsistent file timestamps. Thus cache lenient prevents you from unwanted restarting of cached processes.</strong></p>
</section>
<section id="directives">
<h3>Directives<a class="headerlink" href="#directives" title="Permalink to this heading"></a></h3>
<p>The <a class="reference external" href="https://www.nextflow.io/docs/latest/process.html#directives">directives</a> are declaration blocks that can provide optional settings for a process.</p>
<p>For example, they can affect the way a process stages in and out the input and output files (<a class="reference external" href="https://www.nextflow.io/docs/latest/process.html#stageinmode">stageInMode</a> and <a class="reference external" href="https://www.nextflow.io/docs/latest/process.html#stageoutmode">stageOutMode</a>),  or specify a particular resource such as the number of <a class="reference external" href="https://www.nextflow.io/docs/latest/process.html#cpus">cpus</a> , the <a class="reference external" href="https://www.nextflow.io/docs/latest/process.html#memory">memory</a>, and the <a class="reference external" href="https://www.nextflow.io/docs/latest/process.html#time">time</a></p>
<p>Here an example:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="n">process</span><span class="w"> </span><span class="n">my_process</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">time</span><span class="w"> </span><span class="s1">&#39;1h&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">memory</span><span class="w"> </span><span class="s1">&#39;2 GB&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">cpus</span><span class="w"> </span><span class="mi">8</span><span class="w"></span>

<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    Some execution</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
<p>We can also indicate what to do if a process fails.</p>
<p>The default is to stop the pipeline and to raise an error. But we can also skip the process using the <a class="reference external" href="https://www.nextflow.io/docs/latest/process.html#errorstrategy">errorStrategy</a> directive:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="n">process</span><span class="w"> </span><span class="n">my_process</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">time</span><span class="w"> </span><span class="s1">&#39;1h&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">memory</span><span class="w"> </span><span class="s1">&#39;2 GB&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">cpus</span><span class="w"> </span><span class="mi">8</span><span class="w"></span>
<span class="w">    </span><span class="n">errorStrategy</span><span class="w"> </span><span class="s1">&#39;ignore&#39;</span><span class="w"></span>

<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    Some execution</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
<p>or retry a number of times changing the available memory or the maximum execution time, using the foolowing directives:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="n">process</span><span class="w"> </span><span class="n">my_process</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">memory</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="mi">1</span><span class="o">.</span><span class="na">GB</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">task</span><span class="o">.</span><span class="na">attempt</span><span class="w"> </span><span class="o">}</span><span class="w"></span>
<span class="w">    </span><span class="n">time</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="mi">1</span><span class="o">.</span><span class="na">hour</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">task</span><span class="o">.</span><span class="na">attempt</span><span class="w"> </span><span class="o">}</span><span class="w"></span>
<span class="w">    </span><span class="n">errorStrategy</span><span class="w"> </span><span class="s1">&#39;retry&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">maxRetries</span><span class="w"> </span><span class="mi">3</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
</section>
<section id="id1">
<h3>EXERCISE<a class="headerlink" href="#id1" title="Permalink to this heading"></a></h3>
<p>Make the previous pipeline resilient to the process failing and save the results so the process execution would be skipped when the pipeline is launched again.</p>
<p>First, make the process <cite>reverseSequence</cite> to fail by introducing a typo in the command line, then add the directive to the process.</p>
<details>
<summary><a>Solution</a></summary><p>The solution is at <strong>sol3.nf</strong>. In particular the change is here:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="cp">#!/usr/bin/env nextflow</span><span class="w"></span>

<span class="n">nextflow</span><span class="o">.</span><span class="na">enable</span><span class="o">.</span><span class="na">dsl</span><span class="o">=</span><span class="mi">2</span><span class="w"></span>

<span class="c1">// this can be overridden by using --inputfile OTHERFILENAME</span><span class="w"></span>
<span class="n">params</span><span class="o">.</span><span class="na">inputfile</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;$baseDir/../../testdata/test.fa&quot;</span><span class="w"></span>

<span class="c1">// the &quot;file method&quot; returns a file system object given a file path string</span><span class="w"></span>
<span class="n">sequences_file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">inputfile</span><span class="o">)</span><span class="w"></span>

<span class="c1">// check if the file exists</span><span class="w"></span>
<span class="k">if</span><span class="o">(</span><span class="w"> </span><span class="o">!</span><span class="n">sequences_file</span><span class="o">.</span><span class="na">exists</span><span class="o">()</span><span class="w"> </span><span class="o">)</span><span class="w"> </span><span class="n">exit</span><span class="w"> </span><span class="mi">1</span><span class="o">,</span><span class="w"> </span><span class="s2">&quot;Missing genome file: ${sequences_file}&quot;</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Process 1 for splitting a fasta file in multiple files</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="n">splitSequences</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">sequencesFile</span><span class="w"></span>

<span class="w">    </span><span class="nl">output:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="o">(</span><span class="s1">&#39;seq_*&#39;</span><span class="o">)</span><span class="w"></span>

<span class="w">    </span><span class="c1">// simple awk command</span><span class="w"></span>
<span class="w">    </span><span class="nl">script:</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    awk &#39;/^&gt;/{f=&quot;seq_&quot;++d} {print &gt; f}&#39; &lt; ${sequencesFile}</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="hll"><span class="cm">/*</span>
</span><span class="hll"><span class="cm">* Broken process</span>
</span><span class="hll"><span class="cm">*/</span><span class="w"></span>
</span><span class="hll">
</span><span class="hll"><span class="w"> </span><span class="n">process</span><span class="w"> </span><span class="n">reverseSequence</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
</span><span class="hll">
</span><span class="hll"><span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="s2">&quot;${seq}&quot;</span><span class="w"> </span><span class="o">}</span><span class="w"></span>
</span><span class="hll">
</span><span class="hll"><span class="w">    </span><span class="n">errorStrategy</span><span class="w"> </span><span class="s1">&#39;ignore&#39;</span><span class="w"></span>
</span><span class="hll">
</span><span class="hll"><span class="w">    </span><span class="nl">input:</span><span class="w"></span>
</span><span class="hll"><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">seq</span><span class="w"></span>
</span><span class="hll">
</span><span class="hll"><span class="w">    </span><span class="nl">output:</span><span class="w"></span>
</span><span class="hll"><span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;all.rev&quot;</span><span class="w"></span>
</span><span class="hll">
</span><span class="hll"><span class="w">    </span><span class="nl">script:</span><span class="w"></span>
</span><span class="hll"><span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
</span><span class="hll"><span class="s2">    cat ${seq} | AAAAAAA &#39;{if (\$1~&quot;&gt;&quot;) {print \$0} else system(&quot;echo &quot; \$0 &quot; |rev&quot;)}&#39; &gt; all.rev</span>
</span><span class="hll"><span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
</span><span class="hll"><span class="o">}</span><span class="w"></span>
</span><span class="hll">
</span><span class="n">workflow</span><span class="w"> </span><span class="n">flow1</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="nl">take:</span><span class="w"> </span><span class="n">sequences</span><span class="w"></span>

<span class="w">    </span><span class="nl">main:</span><span class="w"></span>
<span class="w">    </span><span class="n">splitted_seq</span><span class="w">        </span><span class="o">=</span><span class="w"> </span><span class="n">splitSequences</span><span class="o">(</span><span class="n">sequences</span><span class="o">)</span><span class="w"></span>
<span class="w">    </span><span class="n">rev_single_seq</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="n">reverseSequence</span><span class="o">(</span><span class="n">splitted_seq</span><span class="o">)</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="n">workflow</span><span class="w"> </span><span class="n">flow2</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="nl">take:</span><span class="w"> </span><span class="n">sequences</span><span class="w"></span>

<span class="w">    </span><span class="nl">main:</span><span class="w"></span>
<span class="w">    </span><span class="n">splitted_seq</span><span class="w">        </span><span class="o">=</span><span class="w"> </span><span class="n">splitSequences</span><span class="o">(</span><span class="n">sequences</span><span class="o">).</span><span class="na">flatten</span><span class="o">()</span><span class="w"></span>
<span class="w">    </span><span class="n">rev_single_seq</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="n">reverseSequence</span><span class="o">(</span><span class="n">splitted_seq</span><span class="o">)</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="n">workflow</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">   </span><span class="n">flow1</span><span class="o">(</span><span class="n">sequences_file</span><span class="o">)</span><span class="w"></span>
<span class="w">   </span><span class="n">flow2</span><span class="o">(</span><span class="n">sequences_file</span><span class="o">)</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
</details><div class="line-block">
<div class="line"><br /></div>
<div class="line"><br /></div>
</div>
<p>Write the first workflow using pipes. Nextflow DLS2 allows you to use pipes for connecting channels via input / output.</p>
<p>See the <a class="reference external" href="https://www.nextflow.io/docs/latest/dsl2.html#pipes">documentation on pipes</a>.</p>
<details>
<summary><a>Solution</a></summary><p>The solution is at <strong>sol4.nf</strong>. Here is the change:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="cp">#!/usr/bin/env nextflow</span><span class="w"></span>

<span class="n">nextflow</span><span class="o">.</span><span class="na">enable</span><span class="o">.</span><span class="na">dsl</span><span class="o">=</span><span class="mi">2</span><span class="w"></span>

<span class="c1">// this can be overridden by using --inputfile OTHERFILENAME</span><span class="w"></span>
<span class="n">params</span><span class="o">.</span><span class="na">inputfile</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;$baseDir/../../testdata/test.fa&quot;</span><span class="w"></span>

<span class="c1">// the &quot;file method&quot; returns a file system object given a file path string</span><span class="w"></span>
<span class="n">sequences_file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">inputfile</span><span class="o">)</span><span class="w"></span>

<span class="c1">// check if the file exists</span><span class="w"></span>
<span class="k">if</span><span class="o">(</span><span class="w"> </span><span class="o">!</span><span class="n">sequences_file</span><span class="o">.</span><span class="na">exists</span><span class="o">()</span><span class="w"> </span><span class="o">)</span><span class="w"> </span><span class="n">exit</span><span class="w"> </span><span class="mi">1</span><span class="o">,</span><span class="w"> </span><span class="s2">&quot;Missing genome file: ${sequences_file}&quot;</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Process 1 for splitting a fasta file in multiple files</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="n">splitSequences</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">sequencesFile</span><span class="w"></span>

<span class="w">    </span><span class="nl">output:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="o">(</span><span class="s1">&#39;seq_*&#39;</span><span class="o">)</span><span class="w"></span>

<span class="w">    </span><span class="c1">// simple awk command</span><span class="w"></span>
<span class="w">    </span><span class="nl">script:</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    awk &#39;/^&gt;/{f=&quot;seq_&quot;++d} {print &gt; f}&#39; &lt; ${sequencesFile}</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Process 2 for reversing the sequences</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="n">reverseSequence</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="s2">&quot;${seq}&quot;</span><span class="w"> </span><span class="o">}</span><span class="w"></span>

<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">seq</span><span class="w"></span>

<span class="w">    </span><span class="nl">output:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;all.rev&quot;</span><span class="w"></span>

<span class="w">    </span><span class="nl">script:</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    	cat ${seq} | awk &#39;{if (\$1~&quot;&gt;&quot;) {print \$0} else system(&quot;echo &quot; \$0 &quot; |rev&quot;)}&#39; &gt; all.rev</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="hll"><span class="n">workflow</span><span class="w"> </span><span class="n">flow1</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
</span><span class="hll"><span class="w">    </span><span class="nl">take:</span><span class="w"> </span><span class="n">sequences</span><span class="w"></span>
</span><span class="hll">
</span><span class="hll"><span class="w">    </span><span class="nl">main:</span><span class="w"></span>
</span><span class="hll"><span class="w">    </span><span class="n">splitSequences</span><span class="o">(</span><span class="n">sequences</span><span class="o">)</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">reverseSequence</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">view</span><span class="o">()</span><span class="w"></span>
</span><span class="hll"><span class="o">}</span><span class="w"></span>
</span>
<span class="n">workflow</span><span class="w"> </span><span class="n">flow2</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="nl">take:</span><span class="w"> </span><span class="n">sequences</span><span class="w"></span>

<span class="w">    </span><span class="nl">main:</span><span class="w"></span>
<span class="w">    </span><span class="n">splitted_seq</span><span class="w">        </span><span class="o">=</span><span class="w"> </span><span class="n">splitSequences</span><span class="o">(</span><span class="n">sequences</span><span class="o">).</span><span class="na">flatten</span><span class="o">()</span><span class="w"></span>
<span class="w">    </span><span class="n">rev_single_seq</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="n">reverseSequence</span><span class="o">(</span><span class="n">splitted_seq</span><span class="o">)</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="n">workflow</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">   </span><span class="n">flow1</span><span class="o">(</span><span class="n">sequences_file</span><span class="o">)</span><span class="w"></span>
<span class="w">   </span><span class="n">flow2</span><span class="o">(</span><span class="n">sequences_file</span><span class="o">)</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
</details><div class="line-block">
<div class="line"><br /></div>
<div class="line"><br /></div>
</div>
</section>
</section>
<section id="decoupling-resources-parameters-and-nextflow-script">
<h2>Decoupling resources, parameters and nextflow script<a class="headerlink" href="#decoupling-resources-parameters-and-nextflow-script" title="Permalink to this heading"></a></h2>
<p>When making a complex pipelines it is convenient to keep the definition of resources needed, the default parameters and the main script separately from each other.
This can be achieved using two additional files:</p>
<ul class="simple">
<li><p>nextflow.config</p></li>
<li><p>params.config</p></li>
</ul>
<p>The <strong>nextflow.config</strong> file allows to indicate resources needed for each class of processes.
This is achieved labeling processes in the nextflow.config file:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="n">includeConfig</span><span class="w"> </span><span class="s2">&quot;$baseDir/params.config&quot;</span><span class="w"></span>

<span class="n">process</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">memory</span><span class="o">=</span><span class="s1">&#39;0.6G&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">cpus</span><span class="o">=</span><span class="s1">&#39;1&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">time</span><span class="o">=</span><span class="s1">&#39;6h&#39;</span><span class="w"></span>

<span class="w">    </span><span class="nl">withLabel:</span><span class="w"> </span><span class="s1">&#39;onecpu&#39;</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">        </span><span class="n">memory</span><span class="o">=</span><span class="s1">&#39;0.6G&#39;</span><span class="w"></span>
<span class="w">        </span><span class="n">cpus</span><span class="o">=</span><span class="s1">&#39;1&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">}</span><span class="w"></span>

<span class="w">    </span><span class="nl">withLabel:</span><span class="w"> </span><span class="s1">&#39;bigmem&#39;</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">        </span><span class="n">memory</span><span class="o">=</span><span class="s1">&#39;0.7G&#39;</span><span class="w"></span>
<span class="w">        </span><span class="n">cpus</span><span class="o">=</span><span class="s1">&#39;1&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">}</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="n">process</span><span class="o">.</span><span class="na">container</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;biocorecrg/c4lwg-2018:latest&#39;</span><span class="w"></span>
<span class="n">singularity</span><span class="o">.</span><span class="na">cacheDir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;$baseDir/singularity&quot;</span><span class="w"></span>
</pre></div>
</div>
<p>The first row indicates to use the information stored in the <strong>params.config</strong> file (described later). Then follows the definition of the default resources for a process:</p>
<p>Then we specify resources needed for a class of processes labeled <strong>bigmem</strong> (i.e., the default options will be overridden for these processes):</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="nl">withLabel:</span><span class="w"> </span><span class="s1">&#39;bigmem&#39;</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">memory</span><span class="o">=</span><span class="s1">&#39;0.7G&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">cpus</span><span class="o">=</span><span class="s1">&#39;1&#39;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
<p>In the script <strong>/test2/test2.nf file</strong>, there are two processes to run two programs:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">fastQC</a> - a tool that calculates a number of quality control metrics on single fastq files;</p></li>
<li><p><a class="reference external" href="https://multiqc.info/">multiQC</a> - an aggregator of results from bioinformatics tools and samples for generating a single html report.</p></li>
</ul>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="cp">#!/usr/bin/env nextflow</span><span class="w"></span>


<span class="cm">/* </span>
<span class="cm"> * This code enables the new dsl of Nextflow. </span>
<span class="cm"> */</span><span class="w"></span>

<span class="n">nextflow</span><span class="o">.</span><span class="na">enable</span><span class="o">.</span><span class="na">dsl</span><span class="o">=</span><span class="mi">2</span><span class="w"></span>


<span class="cm">/* </span>
<span class="cm"> * NextFlow test pipe</span>
<span class="cm"> * @authors</span>
<span class="cm"> * Luca Cozzuto &lt;lucacozzuto@gmail.com&gt;</span>
<span class="cm"> * </span>
<span class="cm"> */</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Input parameters: read pairs</span>
<span class="cm"> * Params are stored in the params.config file</span>
<span class="cm"> */</span><span class="w"></span>

<span class="n">version</span><span class="w">                 </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;1.0&quot;</span><span class="w"></span>
<span class="c1">// this prevents a warning of undefined parameter</span><span class="w"></span>
<span class="n">params</span><span class="o">.</span><span class="na">help</span><span class="w">             </span><span class="o">=</span><span class="w"> </span><span class="kc">false</span><span class="w"></span>

<span class="c1">// this prints the input parameters</span><span class="w"></span>
<span class="n">log</span><span class="o">.</span><span class="na">info</span><span class="w"> </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">BIOCORE@CRG - N F TESTPIPE  ~  version ${version}</span>
<span class="s2">=============================================</span>
<span class="s2">reads                           : ${params.reads}</span>
<span class="s2">&quot;&quot;&quot;</span><span class="w"></span>

<span class="c1">// this prints the help in case you use --help parameter in the command line and it stops the pipeline</span><span class="w"></span>
<span class="k">if</span><span class="w"> </span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">help</span><span class="o">)</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">log</span><span class="o">.</span><span class="na">info</span><span class="w"> </span><span class="s1">&#39;This is the Biocore\&#39;s NF test pipeline&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">log</span><span class="o">.</span><span class="na">info</span><span class="w"> </span><span class="s1">&#39;Enjoy!&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">log</span><span class="o">.</span><span class="na">info</span><span class="w"> </span><span class="s1">&#39;\n&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">exit</span><span class="w"> </span><span class="mi">1</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Defining the output folders.</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">fastqcOutputFolder</span><span class="w">    </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;ouptut_fastqc&quot;</span><span class="w"></span>
<span class="n">multiqcOutputFolder</span><span class="w">   </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;ouptut_multiQC&quot;</span><span class="w"></span>


<span class="cm">/* Reading the file list and creating a &quot;Channel&quot;: a queue that connects different channels.</span>
<span class="cm"> * The queue is consumed by channels, so you cannot re-use a channel for different processes. </span>
<span class="cm"> * If you need the same data for different processes you need to make more channels.</span>
<span class="cm"> */</span><span class="w"></span>
<span class="w"> </span>
<span class="n">Channel</span><span class="w"></span>
<span class="w">    </span><span class="o">.</span><span class="na">fromPath</span><span class="o">(</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="w"> </span><span class="o">)</span><span class="w">  											 </span><span class="c1">// read the files indicated by the wildcard                            </span><span class="w"></span>
<span class="w">    </span><span class="o">.</span><span class="na">ifEmpty</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="n">error</span><span class="w"> </span><span class="s2">&quot;Cannot find any reads matching: ${params.reads}&quot;</span><span class="w"> </span><span class="o">}</span><span class="w"> </span><span class="c1">// if empty, complains</span><span class="w"></span>
<span class="w">    </span><span class="o">.</span><span class="na">set</span><span class="w"> </span><span class="o">{</span><span class="n">reads_for_fastqc</span><span class="o">}</span><span class="w"> 											 </span><span class="c1">// make the channel &quot;reads_for_fastqc&quot;</span><span class="w"></span>


<span class="cm">/*</span>
<span class="cm"> * Process 1. Run FastQC on raw data. A process is the element for executing scripts / programs etc.</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="n">fastQC</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">publishDir</span><span class="w"> </span><span class="n">fastqcOutputFolder</span><span class="w">  			</span><span class="c1">// where (and whether) to publish the results</span><span class="w"></span>
<span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="s2">&quot;${reads}&quot;</span><span class="w"> </span><span class="o">}</span><span class="w">  							</span><span class="c1">// during the execution prints the indicated variable for follow-up</span><span class="w"></span>
<span class="hll"><span class="w">    </span><span class="n">label</span><span class="w"> </span><span class="s1">&#39;big_mem&#39;</span><span class="w"> </span>
</span>
<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">reads</span><span class="w">   							</span><span class="c1">// it defines the input of the process. It sets values from a channel</span><span class="w"></span>

<span class="w">    </span><span class="nl">output:</span><span class="w">									</span><span class="c1">// It defines the output of the process (i.e. files) and send to a new channel</span><span class="w"></span>
<span class="w">   	</span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;*_fastqc.*&quot;</span><span class="w"></span>

<span class="w">    </span><span class="nl">script:</span><span class="w">									</span><span class="c1">// here you have the execution of the script / program. Basically is the command line</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        fastqc ${reads} </span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Process 2. Run multiQC on fastQC results</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="n">multiQC</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">publishDir</span><span class="w"> </span><span class="n">multiqcOutputFolder</span><span class="o">,</span><span class="w"> </span><span class="nl">mode:</span><span class="w"> </span><span class="s1">&#39;copy&#39;</span><span class="w"> 	</span><span class="c1">// this time do not link but copy the output file</span><span class="w"></span>

<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="o">(</span><span class="n">inputfiles</span><span class="o">)</span><span class="w"></span>

<span class="w">    </span><span class="nl">output:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="o">(</span><span class="s2">&quot;multiqc_report.html&quot;</span><span class="o">)</span><span class="w"> 					</span><span class="c1">// do not send the results to any channel</span><span class="w"></span>

<span class="w">    </span><span class="nl">script:</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    multiqc .</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="n">workflow</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">	</span><span class="n">fastqc_out</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">fastQC</span><span class="o">(</span><span class="n">reads_for_fastqc</span><span class="o">)</span><span class="w"></span>
<span class="w">	</span><span class="n">multiQC</span><span class="o">(</span><span class="n">fastqc_out</span><span class="o">.</span><span class="na">collect</span><span class="o">())</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>


<span class="n">workflow</span><span class="o">.</span><span class="na">onComplete</span><span class="w"> </span><span class="o">{</span><span class="w"> </span>
<span class="w">	</span><span class="n">println</span><span class="w"> </span><span class="o">(</span><span class="w"> </span><span class="n">workflow</span><span class="o">.</span><span class="na">success</span><span class="w"> </span><span class="o">?</span><span class="w"> </span><span class="s2">&quot;\nDone! Open the following report in your browser --&gt; ${multiqcOutputFolder}/multiqc_report.html\n&quot;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;Oops .. something went wrong&quot;</span><span class="w"> </span><span class="o">)</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
<p>You can see that the process <strong>fastQC</strong> is labeled ‘bigmem’.</p>
<p>The last two rows of the config file indicate which containers to use.
In this example, – and by default, if the repository is not specified, – a container is pulled from the DockerHub.
In case of using a singularity container, you can indicate where to store the local image using the <strong>singularity.cacheDir</strong> option:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="n">process</span><span class="o">.</span><span class="na">container</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;biocorecrg/c4lwg-2018:latest&#39;</span><span class="w"></span>
<span class="n">singularity</span><span class="o">.</span><span class="na">cacheDir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;$baseDir/singularity&quot;</span><span class="w"></span>
</pre></div>
</div>
<p>Let’s now launch the script <strong>test2.nf</strong>.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">     cd test2;</span>
<span class="go">     nextflow run test2.nf</span>

<span class="go">     N E X T F L O W  ~  version 20.07.1</span>
<span class="go">     Launching `test2.nf` [distracted_edison] - revision: e3a80b15a2</span>
<span class="go">     BIOCORE@CRG - N F TESTPIPE  ~  version 1.0</span>
<span class="go">     =============================================</span>
<span class="go">     reads                           : /home/ec2-user/git/CRG_Nextflow_Jun_2022/nextflow/nextflow/test2/../testdata/*.fastq.gz</span>
<span class="go">     executor &gt;  local (2)</span>
<span class="go">     [df/2c45f2] process &gt; fastQC (B7_input_s_chr19.fastq.gz) [  0%] 0 of 2</span>
<span class="go">     [-        ] process &gt; multiQC                            -</span>
<span class="go">     Error executing process &gt; &#39;fastQC (B7_H3K4me1_s_chr19.fastq.gz)&#39;</span>

<span class="go">     Caused by:</span>
<span class="go">       Process `fastQC (B7_H3K4me1_s_chr19.fastq.gz)` terminated with an error exit status (127)</span>

<span class="go">     Command executed:</span>

<span class="go">       fastqc B7_H3K4me1_s_chr19.fastq.gz</span>

<span class="go">     Command exit status:</span>
<span class="go">       127</span>

<span class="go">     executor &gt;  local (2)</span>
<span class="go">     [df/2c45f2] process &gt; fastQC (B7_input_s_chr19.fastq.gz) [100%] 2 of 2, failed: 2 ✘</span>
<span class="go">     [-        ] process &gt; multiQC                            -</span>
<span class="go">     Error executing process &gt; &#39;fastQC (B7_H3K4me1_s_chr19.fastq.gz)&#39;</span>

<span class="go">     Caused by:</span>
<span class="go">       Process `fastQC (B7_H3K4me1_s_chr19.fastq.gz)` terminated with an error exit status (127)</span>

<span class="go">     Command executed:</span>

<span class="go">       fastqc B7_H3K4me1_s_chr19.fastq.gz</span>

<span class="go">     Command exit status:</span>
<span class="go">       127</span>

<span class="go">     Command output:</span>
<span class="go">       (empty)</span>

<span class="hll"><span class="go">     Command error:</span>
</span><span class="hll"><span class="go">       .command.sh: line 2: fastqc: command not found</span>
</span>
<span class="go">     Work dir:</span>
<span class="go">       /home/ec2-user/git/CRG_Nextflow_Jun_2022/nextflow/nextflow/test2/work/c5/18e76b2e6ffd64aac2b52e69bedef3</span>

<span class="go">     Tip: when you have fixed the problem you can continue the execution adding the option `-resume` to the run command line</span>
</pre></div>
</div>
<p>We will get a number of errors since no executable is found in our environment / path. This is because the executables are stored in our docker image and we have to tell Nextflow to use the docker image, using the <cite>-with-docker</cite> parameter.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="hll"><span class="go">     nextflow run test2.nf -with-docker</span>
</span>
<span class="go">     N E X T F L O W  ~  version 20.07.1</span>
<span class="go">     Launching `test2.nf` [boring_hamilton] - revision: e3a80b15a2</span>
<span class="go">     BIOCORE@CRG - N F TESTPIPE  ~  version 1.0</span>
<span class="go">     =============================================</span>
<span class="go">     reads                           : /home/ec2-user/git/CRG_Nextflow_Jun_2022/nextflow/nextflow/test2/../testdata/*.fastq.gz</span>
<span class="go">     executor &gt;  local (3)</span>
<span class="go">     [22/b437be] process &gt; fastQC (B7_H3K4me1_s_chr19.fastq.gz) [100%] 2 of 2 ✔</span>
<span class="go">     [1a/cfe63b] process &gt; multiQC                              [  0%] 0 of 1</span>
<span class="go">     executor &gt;  local (3)</span>
<span class="go">     [22/b437be] process &gt; fastQC (B7_H3K4me1_s_chr19.fastq.gz) [100%] 2 of 2 ✔</span>
<span class="go">     [1a/cfe63b] process &gt; multiQC                              [100%] 1 of 1 ✔</span>
</pre></div>
</div>
<p>This time it worked because Nextflow used the image specified in the <strong>nextflow.config</strong> file and containing the executables.</p>
<p>Now let’s take a look at the <strong>params.config</strong> file:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="n">params</span><span class="w"> </span><span class="o">{</span><span class="w"></span>

<span class="w">	</span><span class="n">reads</span><span class="w">		</span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;$baseDir/../../testdata/*.fastq.gz&quot;</span><span class="w"></span>
<span class="w">	</span><span class="n">email</span><span class="w">		</span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;myemail@google.com&quot;</span><span class="w"></span>

<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
<p>As you can see, we indicated two pipeline parameters, <cite>reads</cite> and <cite>email</cite>; when running the pipeline, they can be overridden using <cite>--reads</cite> and <cite>--email</cite>.</p>
<p>Now, let’s examine the folders generated by the pipeline.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">ls  work/2a/22e3df887b1b5ac8af4f9cd0d88ac5/</span>

<span class="go">total 0</span>
<span class="go">drwxrwxr-x 3 ec2-user ec2-user  26 Apr 23 13:52 .</span>
<span class="go">drwxr-xr-x 2 root     root     136 Apr 23 13:51 multiqc_data</span>
<span class="go">drwxrwxr-x 3 ec2-user ec2-user  44 Apr 23 13:51 ..</span>
</pre></div>
</div>
<p>We observe that Docker runs as “root”. This can be problematic and generates security issues. To avoid this we can add this line of code within the process section of the config file:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="n">containerOptions</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="n">workflow</span><span class="o">.</span><span class="na">containerEngine</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">&quot;docker&quot;</span><span class="w"> </span><span class="o">?</span><span class="w"> </span><span class="s1">&#39;-u $(id -u):$(id -g)&#39;</span><span class="o">:</span><span class="w"> </span><span class="kc">null</span><span class="o">}</span><span class="w"></span>
</pre></div>
</div>
<p>This will tell Nextflow that if it is run with Docker, it has to produce files that belong to a user rather than the root.</p>
<section id="publishing-final-results">
<h3>Publishing final results<a class="headerlink" href="#publishing-final-results" title="Permalink to this heading"></a></h3>
<p>The script <strong>test2.nf</strong> generates two new folders, <strong>output_fastqc</strong> and <strong>output_multiQC</strong>, that contain the result of the pipeline output.
We can indicate which process and output can be considered the final output of the pipeline using the <strong>publishDir</strong> directive that has to be specified at the beginning of a process.</p>
<p>In our pipeline we define these folders here:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="cp">#!/usr/bin/env nextflow</span><span class="w"></span>


<span class="cm">/* </span>
<span class="cm"> * This code enables the new dsl of Nextflow. </span>
<span class="cm"> */</span><span class="w"></span>

<span class="n">nextflow</span><span class="o">.</span><span class="na">enable</span><span class="o">.</span><span class="na">dsl</span><span class="o">=</span><span class="mi">2</span><span class="w"></span>


<span class="cm">/* </span>
<span class="cm"> * NextFlow test pipe</span>
<span class="cm"> * @authors</span>
<span class="cm"> * Luca Cozzuto &lt;lucacozzuto@gmail.com&gt;</span>
<span class="cm"> * </span>
<span class="cm"> */</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Input parameters: read pairs</span>
<span class="cm"> * Params are stored in the params.config file</span>
<span class="cm"> */</span><span class="w"></span>

<span class="n">version</span><span class="w">                 </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;1.0&quot;</span><span class="w"></span>
<span class="c1">// this prevents a warning of undefined parameter</span><span class="w"></span>
<span class="n">params</span><span class="o">.</span><span class="na">help</span><span class="w">             </span><span class="o">=</span><span class="w"> </span><span class="kc">false</span><span class="w"></span>

<span class="c1">// this prints the input parameters</span><span class="w"></span>
<span class="n">log</span><span class="o">.</span><span class="na">info</span><span class="w"> </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">BIOCORE@CRG - N F TESTPIPE  ~  version ${version}</span>
<span class="s2">=============================================</span>
<span class="s2">reads                           : ${params.reads}</span>
<span class="s2">&quot;&quot;&quot;</span><span class="w"></span>

<span class="c1">// this prints the help in case you use --help parameter in the command line and it stops the pipeline</span><span class="w"></span>
<span class="k">if</span><span class="w"> </span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">help</span><span class="o">)</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">log</span><span class="o">.</span><span class="na">info</span><span class="w"> </span><span class="s1">&#39;This is the Biocore\&#39;s NF test pipeline&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">log</span><span class="o">.</span><span class="na">info</span><span class="w"> </span><span class="s1">&#39;Enjoy!&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">log</span><span class="o">.</span><span class="na">info</span><span class="w"> </span><span class="s1">&#39;\n&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">exit</span><span class="w"> </span><span class="mi">1</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="hll"><span class="cm">/*</span>
</span><span class="hll"><span class="cm"> * Defining the output folders.</span>
</span><span class="hll"><span class="cm"> */</span><span class="w"></span>
</span><span class="hll"><span class="n">fastqcOutputFolder</span><span class="w">    </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;ouptut_fastqc&quot;</span><span class="w"></span>
</span><span class="hll"><span class="n">multiqcOutputFolder</span><span class="w">   </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;ouptut_multiQC&quot;</span><span class="w"></span>
</span>

<span class="cm">/* Reading the file list and creating a &quot;Channel&quot;: a queue that connects different channels.</span>
<span class="cm"> * The queue is consumed by channels, so you cannot re-use a channel for different processes. </span>
<span class="cm"> * If you need the same data for different processes you need to make more channels.</span>
<span class="cm"> */</span><span class="w"></span>
<span class="w"> </span>
<span class="n">Channel</span><span class="w"></span>
<span class="w">    </span><span class="o">.</span><span class="na">fromPath</span><span class="o">(</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="w"> </span><span class="o">)</span><span class="w">  											 </span><span class="c1">// read the files indicated by the wildcard                            </span><span class="w"></span>
<span class="w">    </span><span class="o">.</span><span class="na">ifEmpty</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="n">error</span><span class="w"> </span><span class="s2">&quot;Cannot find any reads matching: ${params.reads}&quot;</span><span class="w"> </span><span class="o">}</span><span class="w"> </span><span class="c1">// if empty, complains</span><span class="w"></span>
<span class="w">    </span><span class="o">.</span><span class="na">set</span><span class="w"> </span><span class="o">{</span><span class="n">reads_for_fastqc</span><span class="o">}</span><span class="w"> 											 </span><span class="c1">// make the channel &quot;reads_for_fastqc&quot;</span><span class="w"></span>


<span class="cm">/*</span>
<span class="cm"> * Process 1. Run FastQC on raw data. A process is the element for executing scripts / programs etc.</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="n">fastQC</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="hll"><span class="w">    </span><span class="n">publishDir</span><span class="w"> </span><span class="n">fastqcOutputFolder</span><span class="w">  			</span><span class="c1">// where (and whether) to publish the results</span><span class="w"></span>
</span><span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="s2">&quot;${reads}&quot;</span><span class="w"> </span><span class="o">}</span><span class="w">  							</span><span class="c1">// during the execution prints the indicated variable for follow-up</span><span class="w"></span>
<span class="w">    </span><span class="n">label</span><span class="w"> </span><span class="s1">&#39;big_mem&#39;</span><span class="w"> </span>

<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">reads</span><span class="w">   							</span><span class="c1">// it defines the input of the process. It sets values from a channel</span><span class="w"></span>

<span class="w">    </span><span class="nl">output:</span><span class="w">									</span><span class="c1">// It defines the output of the process (i.e. files) and send to a new channel</span><span class="w"></span>
<span class="w">   	</span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;*_fastqc.*&quot;</span><span class="w"></span>

<span class="w">    </span><span class="nl">script:</span><span class="w">									</span><span class="c1">// here you have the execution of the script / program. Basically is the command line</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        fastqc ${reads} </span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Process 2. Run multiQC on fastQC results</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="n">multiQC</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="hll"><span class="w">    </span><span class="n">publishDir</span><span class="w"> </span><span class="n">multiqcOutputFolder</span><span class="o">,</span><span class="w"> </span><span class="nl">mode:</span><span class="w"> </span><span class="s1">&#39;copy&#39;</span><span class="w"> 	</span><span class="c1">// this time do not link but copy the output file</span><span class="w"></span>
</span>
<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="o">(</span><span class="n">inputfiles</span><span class="o">)</span><span class="w"></span>

<span class="w">    </span><span class="nl">output:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="o">(</span><span class="s2">&quot;multiqc_report.html&quot;</span><span class="o">)</span><span class="w"> 					</span><span class="c1">// do not send the results to any channel</span><span class="w"></span>

<span class="w">    </span><span class="nl">script:</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    multiqc .</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="n">workflow</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">	</span><span class="n">fastqc_out</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">fastQC</span><span class="o">(</span><span class="n">reads_for_fastqc</span><span class="o">)</span><span class="w"></span>
<span class="w">	</span><span class="n">multiQC</span><span class="o">(</span><span class="n">fastqc_out</span><span class="o">.</span><span class="na">collect</span><span class="o">())</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>


<span class="n">workflow</span><span class="o">.</span><span class="na">onComplete</span><span class="w"> </span><span class="o">{</span><span class="w"> </span>
<span class="w">	</span><span class="n">println</span><span class="w"> </span><span class="o">(</span><span class="w"> </span><span class="n">workflow</span><span class="o">.</span><span class="na">success</span><span class="w"> </span><span class="o">?</span><span class="w"> </span><span class="s2">&quot;\nDone! Open the following report in your browser --&gt; ${multiqcOutputFolder}/multiqc_report.html\n&quot;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;Oops .. something went wrong&quot;</span><span class="w"> </span><span class="o">)</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
<p>You can see that the default mode to publish the results in Nextflow is <cite>soft linking</cite>. You can change this behaviour specifying the mode as indicated in the <strong>multiQC</strong> process.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>IMPORTANT: You can also “move” the results but this is not suggested for files that will be needed for other processes. This will likely disrupt your pipeline</p>
</div>
<p>To access the output files via the web they can be copied to your <a class="reference external" href="https://docs.aws.amazon.com/AmazonS3/latest/userguide/UsingBucket.html">S3 bucket</a> . Your bucket is mounted in <strong>/mnt</strong>:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">ls /mnt</span>

<span class="go">/mnt/nf-class-bucket-1</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>In this class, each student has its own bucket, with the number correponding to the number of the AWS instance.</p>
</div>
<p>Let’s copy the <strong>multiqc_report.html</strong> file in the S3 bucket and change the privileges:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">cp output_multiQC/multiqc_report.html /mnt/nf-class-bucket-1</span>

<span class="go">sudo chmod 775 /mnt/nf-class-bucket-1/multiqc_report.html</span>
</pre></div>
</div>
<p>Now you will be able to see this html file via the browser (change the bucket number to correspond to your instance):</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">http://nf-class-bucket-1.s3.eu-central-1.amazonaws.com/multiqc_report.html</span>
</pre></div>
</div>
</section>
</section>
<section id="adding-help-section-to-a-pipeline">
<h2>Adding help section to a pipeline<a class="headerlink" href="#adding-help-section-to-a-pipeline" title="Permalink to this heading"></a></h2>
<p>Here we describe another good practice: the use of the <cite>--help</cite> parameter. At the beginning of the pipeline we can write:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="cp">#!/usr/bin/env nextflow</span><span class="w"></span>


<span class="cm">/* </span>
<span class="cm"> * This code enables the new dsl of Nextflow. </span>
<span class="cm"> */</span><span class="w"></span>

<span class="n">nextflow</span><span class="o">.</span><span class="na">enable</span><span class="o">.</span><span class="na">dsl</span><span class="o">=</span><span class="mi">2</span><span class="w"></span>


<span class="cm">/* </span>
<span class="cm"> * NextFlow test pipe</span>
<span class="cm"> * @authors</span>
<span class="cm"> * Luca Cozzuto &lt;lucacozzuto@gmail.com&gt;</span>
<span class="cm"> * </span>
<span class="cm"> */</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Input parameters: read pairs</span>
<span class="cm"> * Params are stored in the params.config file</span>
<span class="cm"> */</span><span class="w"></span>

<span class="n">version</span><span class="w">                 </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;1.0&quot;</span><span class="w"></span>
<span class="c1">// this prevents a warning of undefined parameter</span><span class="w"></span>
<span class="hll"><span class="n">params</span><span class="o">.</span><span class="na">help</span><span class="w">             </span><span class="o">=</span><span class="w"> </span><span class="kc">false</span><span class="w"></span>
</span><span class="hll">
</span><span class="hll"><span class="c1">// this prints the input parameters</span><span class="w"></span>
</span><span class="hll"><span class="n">log</span><span class="o">.</span><span class="na">info</span><span class="w"> </span><span class="s2">&quot;&quot;&quot;</span>
</span><span class="hll"><span class="s2">BIOCORE@CRG - N F TESTPIPE  ~  version ${version}</span>
</span><span class="hll"><span class="s2">=============================================</span>
</span><span class="hll"><span class="s2">reads                           : ${params.reads}</span>
</span><span class="hll"><span class="s2">&quot;&quot;&quot;</span><span class="w"></span>
</span><span class="hll">
</span><span class="hll"><span class="c1">// this prints the help in case you use --help parameter in the command line and it stops the pipeline</span><span class="w"></span>
</span><span class="hll"><span class="k">if</span><span class="w"> </span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">help</span><span class="o">)</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
</span><span class="hll"><span class="w">    </span><span class="n">log</span><span class="o">.</span><span class="na">info</span><span class="w"> </span><span class="s1">&#39;This is the Biocore\&#39;s NF test pipeline&#39;</span><span class="w"></span>
</span><span class="hll"><span class="w">    </span><span class="n">log</span><span class="o">.</span><span class="na">info</span><span class="w"> </span><span class="s1">&#39;Enjoy!&#39;</span><span class="w"></span>
</span><span class="hll"><span class="w">    </span><span class="n">log</span><span class="o">.</span><span class="na">info</span><span class="w"> </span><span class="s1">&#39;\n&#39;</span><span class="w"></span>
</span><span class="hll"><span class="w">    </span><span class="n">exit</span><span class="w"> </span><span class="mi">1</span><span class="w"></span>
</span><span class="hll"><span class="o">}</span><span class="w"></span>
</span>
<span class="cm">/*</span>
<span class="cm"> * Defining the output folders.</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">fastqcOutputFolder</span><span class="w">    </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;ouptut_fastqc&quot;</span><span class="w"></span>
<span class="n">multiqcOutputFolder</span><span class="w">   </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;ouptut_multiQC&quot;</span><span class="w"></span>


<span class="cm">/* Reading the file list and creating a &quot;Channel&quot;: a queue that connects different channels.</span>
<span class="cm"> * The queue is consumed by channels, so you cannot re-use a channel for different processes. </span>
<span class="cm"> * If you need the same data for different processes you need to make more channels.</span>
<span class="cm"> */</span><span class="w"></span>
<span class="w"> </span>
<span class="n">Channel</span><span class="w"></span>
<span class="w">    </span><span class="o">.</span><span class="na">fromPath</span><span class="o">(</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="w"> </span><span class="o">)</span><span class="w">  											 </span><span class="c1">// read the files indicated by the wildcard                            </span><span class="w"></span>
<span class="w">    </span><span class="o">.</span><span class="na">ifEmpty</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="n">error</span><span class="w"> </span><span class="s2">&quot;Cannot find any reads matching: ${params.reads}&quot;</span><span class="w"> </span><span class="o">}</span><span class="w"> </span><span class="c1">// if empty, complains</span><span class="w"></span>
<span class="w">    </span><span class="o">.</span><span class="na">set</span><span class="w"> </span><span class="o">{</span><span class="n">reads_for_fastqc</span><span class="o">}</span><span class="w"> 											 </span><span class="c1">// make the channel &quot;reads_for_fastqc&quot;</span><span class="w"></span>


<span class="cm">/*</span>
<span class="cm"> * Process 1. Run FastQC on raw data. A process is the element for executing scripts / programs etc.</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="n">fastQC</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">publishDir</span><span class="w"> </span><span class="n">fastqcOutputFolder</span><span class="w">  			</span><span class="c1">// where (and whether) to publish the results</span><span class="w"></span>
<span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="s2">&quot;${reads}&quot;</span><span class="w"> </span><span class="o">}</span><span class="w">  							</span><span class="c1">// during the execution prints the indicated variable for follow-up</span><span class="w"></span>
<span class="w">    </span><span class="n">label</span><span class="w"> </span><span class="s1">&#39;big_mem&#39;</span><span class="w"> </span>

<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">reads</span><span class="w">   							</span><span class="c1">// it defines the input of the process. It sets values from a channel</span><span class="w"></span>

<span class="w">    </span><span class="nl">output:</span><span class="w">									</span><span class="c1">// It defines the output of the process (i.e. files) and send to a new channel</span><span class="w"></span>
<span class="w">   	</span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;*_fastqc.*&quot;</span><span class="w"></span>

<span class="w">    </span><span class="nl">script:</span><span class="w">									</span><span class="c1">// here you have the execution of the script / program. Basically is the command line</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        fastqc ${reads} </span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Process 2. Run multiQC on fastQC results</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="n">multiQC</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">publishDir</span><span class="w"> </span><span class="n">multiqcOutputFolder</span><span class="o">,</span><span class="w"> </span><span class="nl">mode:</span><span class="w"> </span><span class="s1">&#39;copy&#39;</span><span class="w"> 	</span><span class="c1">// this time do not link but copy the output file</span><span class="w"></span>

<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="o">(</span><span class="n">inputfiles</span><span class="o">)</span><span class="w"></span>

<span class="w">    </span><span class="nl">output:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="o">(</span><span class="s2">&quot;multiqc_report.html&quot;</span><span class="o">)</span><span class="w"> 					</span><span class="c1">// do not send the results to any channel</span><span class="w"></span>

<span class="w">    </span><span class="nl">script:</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    multiqc .</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="n">workflow</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">	</span><span class="n">fastqc_out</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">fastQC</span><span class="o">(</span><span class="n">reads_for_fastqc</span><span class="o">)</span><span class="w"></span>
<span class="w">	</span><span class="n">multiQC</span><span class="o">(</span><span class="n">fastqc_out</span><span class="o">.</span><span class="na">collect</span><span class="o">())</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>


<span class="n">workflow</span><span class="o">.</span><span class="na">onComplete</span><span class="w"> </span><span class="o">{</span><span class="w"> </span>
<span class="w">	</span><span class="n">println</span><span class="w"> </span><span class="o">(</span><span class="w"> </span><span class="n">workflow</span><span class="o">.</span><span class="na">success</span><span class="w"> </span><span class="o">?</span><span class="w"> </span><span class="s2">&quot;\nDone! Open the following report in your browser --&gt; ${multiqcOutputFolder}/multiqc_report.html\n&quot;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;Oops .. something went wrong&quot;</span><span class="w"> </span><span class="o">)</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
<p>so that launching the pipeline with <cite>--help</cite> will show you just the parameters and the help.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nextflow run test2.nf --help</span>

<span class="go">N E X T F L O W  ~  version 20.07.1</span>
<span class="go">Launching `test2.nf` [mad_elion] - revision: e3a80b15a2</span>
<span class="go">BIOCORE@CRG - N F TESTPIPE  ~  version 1.0</span>
<span class="go">=============================================</span>
<span class="go">reads                           : /home/ec2-user/git/CRG_Nextflow_Jun_2022/nextflow/nextflow/test2/../testdata/*.fastq.gz</span>
<span class="go">This is the Biocore&#39;s NF test pipeline</span>
<span class="go">Enjoy!</span>
</pre></div>
</div>
</section>
<section id="id2">
<h2>EXERCISE<a class="headerlink" href="#id2" title="Permalink to this heading"></a></h2>
<ul class="simple">
<li><p>Look at the very last EXERCISE of the day before. Change the script and the config file using the label for handling failing processes.</p></li>
</ul>
<details>
<summary><a>Solution</a></summary><p>The process should become:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="cp">#!/usr/bin/env nextflow</span><span class="w"></span>

<span class="n">nextflow</span><span class="o">.</span><span class="na">enable</span><span class="o">.</span><span class="na">dsl</span><span class="o">=</span><span class="mi">2</span><span class="w"></span>

<span class="c1">// this can be overridden by using --inputfile OTHERFILENAME</span><span class="w"></span>
<span class="n">params</span><span class="o">.</span><span class="na">inputfile</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;$baseDir/../../../testdata/test.fa&quot;</span><span class="w"></span>

<span class="c1">// the &quot;file method&quot; returns a file system object given a file path string</span><span class="w"></span>
<span class="n">sequences_file</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">file</span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">inputfile</span><span class="o">)</span><span class="w"></span>

<span class="c1">// check if the file exists</span><span class="w"></span>
<span class="k">if</span><span class="o">(</span><span class="w"> </span><span class="o">!</span><span class="n">sequences_file</span><span class="o">.</span><span class="na">exists</span><span class="o">()</span><span class="w"> </span><span class="o">)</span><span class="w"> </span><span class="n">exit</span><span class="w"> </span><span class="mi">1</span><span class="o">,</span><span class="w"> </span><span class="s2">&quot;Missing genome file: ${sequences_file}&quot;</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Process 1 for splitting a fasta file in multiple files</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="n">splitSequences</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">sequencesFile</span><span class="w"></span>

<span class="w">    </span><span class="nl">output:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="o">(</span><span class="s1">&#39;seq_*&#39;</span><span class="o">)</span><span class="w"></span>

<span class="w">    </span><span class="c1">// simple awk command</span><span class="w"></span>
<span class="w">    </span><span class="nl">script:</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    awk &#39;/^&gt;/{f=&quot;seq_&quot;++d} {print &gt; f}&#39; &lt; ${sequencesFile}</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Process 2 for reversing the sequences</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="n">reverseSequence</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="s2">&quot;${seq}&quot;</span><span class="w"> </span><span class="o">}</span><span class="w"></span>

<span class="w">    </span><span class="n">publishDir</span><span class="w"> </span><span class="s2">&quot;output&quot;</span><span class="w"></span>
<span class="hll"><span class="w">    </span><span class="n">label</span><span class="w"> </span><span class="s1">&#39;ignorefail&#39;</span><span class="w"></span>
</span><span class="w">    </span>
<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">seq</span><span class="w"></span>

<span class="w">    </span><span class="nl">output:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;all.rev&quot;</span><span class="w"></span>

<span class="w">    </span><span class="nl">script:</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    	cat ${seq} | awk &#39;{if (\$1~&quot;&gt;&quot;) {print \$0} else system(&quot;echo &quot; \$0 &quot; |rev&quot;)}&#39; &gt; all.rev</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="n">workflow</span><span class="w"> </span><span class="n">flow1</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="nl">take:</span><span class="w"> </span><span class="n">sequences</span><span class="w"></span>

<span class="w">    </span><span class="nl">main:</span><span class="w"></span>
<span class="w">    </span><span class="n">splitted_seq</span><span class="w">        </span><span class="o">=</span><span class="w"> </span><span class="n">splitSequences</span><span class="o">(</span><span class="n">sequences</span><span class="o">)</span><span class="w"></span>
<span class="w">    </span><span class="n">rev_single_seq</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="n">reverseSequence</span><span class="o">(</span><span class="n">splitted_seq</span><span class="o">)</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="n">workflow</span><span class="w"> </span><span class="n">flow2</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="nl">take:</span><span class="w"> </span><span class="n">sequences</span><span class="w"></span>

<span class="w">    </span><span class="nl">main:</span><span class="w"></span>
<span class="w">    </span><span class="n">splitted_seq</span><span class="w">        </span><span class="o">=</span><span class="w"> </span><span class="n">splitSequences</span><span class="o">(</span><span class="n">sequences</span><span class="o">).</span><span class="na">flatten</span><span class="o">()</span><span class="w"></span>
<span class="w">    </span><span class="n">rev_single_seq</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="n">reverseSequence</span><span class="o">(</span><span class="n">splitted_seq</span><span class="o">)</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="n">workflow</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">   </span><span class="n">flow1</span><span class="o">(</span><span class="n">sequences_file</span><span class="o">)</span><span class="w"></span>
<span class="w">   </span><span class="n">flow2</span><span class="o">(</span><span class="n">sequences_file</span><span class="o">)</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
<p>and the nextflow.config file would become:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="n">process</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="nl">withLabel:</span><span class="w"> </span><span class="s1">&#39;ignorefail&#39;</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">        </span><span class="n">errorStrategy</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;ignore&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">}</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
</details><div class="line-block">
<div class="line"><br /></div>
<div class="line"><br /></div>
</div>
<ul class="simple">
<li><p>Now look at <strong>test2.nf</strong>.</p></li>
</ul>
<p>Change this script and the config file using the label for handling failing processes by retrying 3 times and incrementing time.</p>
<p>You can specify a very low time (5, 10 or 15 seconds) for the fastqc process so it would fail at beginning.</p>
<details>
<summary><a>Solution</a></summary><p>The code should become:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="cp">#!/usr/bin/env nextflow</span><span class="w"></span>


<span class="cm">/* </span>
<span class="cm"> * This code enables the new dsl of Nextflow. </span>
<span class="cm"> */</span><span class="w"></span>

<span class="n">nextflow</span><span class="o">.</span><span class="na">enable</span><span class="o">.</span><span class="na">dsl</span><span class="o">=</span><span class="mi">2</span><span class="w"></span>


<span class="cm">/* </span>
<span class="cm"> * NextFlow test pipe</span>
<span class="cm"> * @authors</span>
<span class="cm"> * Luca Cozzuto &lt;lucacozzuto@gmail.com&gt;</span>
<span class="cm"> * </span>
<span class="cm"> */</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Input parameters: read pairs</span>
<span class="cm"> * Params are stored in the params.config file</span>
<span class="cm"> */</span><span class="w"></span>

<span class="n">version</span><span class="w">                 </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;1.0&quot;</span><span class="w"></span>
<span class="c1">// this prevents a warning of undefined parameter</span><span class="w"></span>
<span class="n">params</span><span class="o">.</span><span class="na">help</span><span class="w">             </span><span class="o">=</span><span class="w"> </span><span class="kc">false</span><span class="w"></span>

<span class="c1">// this prints the input parameters</span><span class="w"></span>
<span class="n">log</span><span class="o">.</span><span class="na">info</span><span class="w"> </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">BIOCORE@CRG - N F TESTPIPE  ~  version ${version}</span>
<span class="s2">=============================================</span>
<span class="s2">reads                           : ${params.reads}</span>
<span class="s2">&quot;&quot;&quot;</span><span class="w"></span>

<span class="c1">// this prints the help in case you use --help parameter in the command line and it stops the pipeline</span><span class="w"></span>
<span class="k">if</span><span class="w"> </span><span class="o">(</span><span class="n">params</span><span class="o">.</span><span class="na">help</span><span class="o">)</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">log</span><span class="o">.</span><span class="na">info</span><span class="w"> </span><span class="s1">&#39;This is the Biocore\&#39;s NF test pipeline&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">log</span><span class="o">.</span><span class="na">info</span><span class="w"> </span><span class="s1">&#39;Enjoy!&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">log</span><span class="o">.</span><span class="na">info</span><span class="w"> </span><span class="s1">&#39;\n&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">exit</span><span class="w"> </span><span class="mi">1</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="cm">/*</span>
<span class="cm"> * Defining the output folders.</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">fastqcOutputFolder</span><span class="w">    </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;ouptut_fastqc&quot;</span><span class="w"></span>
<span class="n">multiqcOutputFolder</span><span class="w">   </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;ouptut_multiQC&quot;</span><span class="w"></span>


<span class="cm">/* Reading the file list and creating a &quot;Channel&quot;: a queue that connects different channels.</span>
<span class="cm"> * The queue is consumed by channels, so you cannot re-use a channel for different processes. </span>
<span class="cm"> * If you need the same data for different processes you need to make more channels.</span>
<span class="cm"> */</span><span class="w"></span>
<span class="w"> </span>
<span class="n">Channel</span><span class="w"></span>
<span class="w">    </span><span class="o">.</span><span class="na">fromPath</span><span class="o">(</span><span class="w"> </span><span class="n">params</span><span class="o">.</span><span class="na">reads</span><span class="w"> </span><span class="o">)</span><span class="w">  											 </span><span class="c1">// read the files indicated by the wildcard                            </span><span class="w"></span>
<span class="w">    </span><span class="o">.</span><span class="na">ifEmpty</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="n">error</span><span class="w"> </span><span class="s2">&quot;Cannot find any reads matching: ${params.reads}&quot;</span><span class="w"> </span><span class="o">}</span><span class="w"> </span><span class="c1">// if empty, complains</span><span class="w"></span>
<span class="w">    </span><span class="o">.</span><span class="na">set</span><span class="w"> </span><span class="o">{</span><span class="n">reads_for_fastqc</span><span class="o">}</span><span class="w"> 											 </span><span class="c1">// make the channel &quot;reads_for_fastqc&quot;</span><span class="w"></span>


<span class="cm">/*</span>
<span class="cm"> * Process 1. Run FastQC on raw data. A process is the element for executing scripts / programs etc.</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="n">fastQC</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">publishDir</span><span class="w"> </span><span class="n">fastqcOutputFolder</span><span class="w">  			</span><span class="c1">// where (and whether) to publish the results</span><span class="w"></span>
<span class="w">    </span><span class="n">tag</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="s2">&quot;${reads}&quot;</span><span class="w"> </span><span class="o">}</span><span class="w">  							</span><span class="c1">// during the execution prints the indicated variable for follow-up</span><span class="w"></span>
<span class="hll"><span class="w">    </span><span class="n">label</span><span class="w"> </span><span class="s1">&#39;keep_trying&#39;</span><span class="w"></span>
</span>
<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="n">reads</span><span class="w">   							</span><span class="c1">// it defines the input of the process. It sets values from a channel</span><span class="w"></span>

<span class="w">    </span><span class="nl">output:</span><span class="w">									</span><span class="c1">// It defines the output of the process (i.e. files) and send to a new channel</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="s2">&quot;*_fastqc.*&quot;</span><span class="w"></span>

<span class="w">    </span><span class="nl">script:</span><span class="w">									</span><span class="c1">// here you have the execution of the script / program. Basically is the command line</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        fastqc ${reads} </span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>


<span class="cm">/*</span>
<span class="cm"> * Process 2. Run multiQC on fastQC results</span>
<span class="cm"> */</span><span class="w"></span>
<span class="n">process</span><span class="w"> </span><span class="n">multiQC</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">publishDir</span><span class="w"> </span><span class="n">multiqcOutputFolder</span><span class="o">,</span><span class="w"> </span><span class="nl">mode:</span><span class="w"> </span><span class="s1">&#39;copy&#39;</span><span class="w"> 	</span><span class="c1">// this time do not link but copy the output file</span><span class="w"></span>

<span class="w">    </span><span class="nl">input:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="w"> </span><span class="o">(</span><span class="n">inputfiles</span><span class="o">)</span><span class="w"></span>

<span class="w">    </span><span class="nl">output:</span><span class="w"></span>
<span class="w">    </span><span class="n">path</span><span class="o">(</span><span class="s2">&quot;multiqc_report.html&quot;</span><span class="o">)</span><span class="w"> 					</span><span class="c1">// do not send the results to any channel</span><span class="w"></span>

<span class="w">    </span><span class="nl">script:</span><span class="w"></span>
<span class="w">    </span><span class="s2">&quot;&quot;&quot;</span>
<span class="s2">       multiqc .</span>
<span class="s2">    &quot;&quot;&quot;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>

<span class="n">workflow</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">	</span><span class="n">fastqc_out</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">fastQC</span><span class="o">(</span><span class="n">reads_for_fastqc</span><span class="o">)</span><span class="w"></span>
<span class="w">	</span><span class="n">multiQC</span><span class="o">(</span><span class="n">fastqc_out</span><span class="o">.</span><span class="na">collect</span><span class="o">())</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>


<span class="n">workflow</span><span class="o">.</span><span class="na">onComplete</span><span class="w"> </span><span class="o">{</span><span class="w"> </span>
<span class="w">	</span><span class="n">println</span><span class="w"> </span><span class="o">(</span><span class="w"> </span><span class="n">workflow</span><span class="o">.</span><span class="na">success</span><span class="w"> </span><span class="o">?</span><span class="w"> </span><span class="s2">&quot;\nDone! Open the following report in your browser --&gt; ${multiqcOutputFolder}/multiqc_report.html\n&quot;</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="s2">&quot;Oops .. something went wrong&quot;</span><span class="w"> </span><span class="o">)</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
<p>while the nextflow.config file would be:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="n">includeConfig</span><span class="w"> </span><span class="s2">&quot;$baseDir/params.config&quot;</span><span class="w"></span>

<span class="n">process</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">memory</span><span class="o">=</span><span class="s1">&#39;0.6G&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">cpus</span><span class="o">=</span><span class="s1">&#39;1&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">time</span><span class="o">=</span><span class="s1">&#39;6h&#39;</span><span class="w"></span>

<span class="w">    </span><span class="nl">withLabel:</span><span class="w"> </span><span class="s1">&#39;keep_trying&#39;</span><span class="w"> </span><span class="o">{</span><span class="w"> </span>
<span class="w">        </span><span class="n">time</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">{</span><span class="w"> </span><span class="mi">10</span><span class="o">.</span><span class="na">second</span><span class="w"> </span><span class="o">*</span><span class="w"> </span><span class="n">task</span><span class="o">.</span><span class="na">attempt</span><span class="w"> </span><span class="o">}</span><span class="w"></span>
<span class="w">        </span><span class="n">errorStrategy</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;retry&#39;</span><span class="w"> </span>
<span class="w">        </span><span class="n">maxRetries</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">3</span><span class="w">	</span>
<span class="w">    </span><span class="o">}</span><span class="w"> 	</span>

<span class="o">}</span><span class="w"></span>

<span class="n">process</span><span class="o">.</span><span class="na">container</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;biocorecrg/c4lwg-2018:latest&#39;</span><span class="w"></span>
<span class="n">singularity</span><span class="o">.</span><span class="na">cacheDir</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">&quot;$baseDir/singularity&quot;</span><span class="w"></span>
</pre></div>
</div>
</details><div class="line-block">
<div class="line"><br /></div>
<div class="line"><br /></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="nextflow_1.html" class="btn btn-neutral float-left" title="Nextflow 1" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="nextflow_3.html" class="btn btn-neutral float-right" title="Nextflow 3" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright .</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>